File: apps/ios/Tests/CrossyAPITests/PaginationTests.swift
import Foundation
import XCTest

import CrossyAPI
import CrossyProtocol

// Cursor pagination for the two list endpoints (PROTOCOL.md section 12): `limit`
// clamped server-side to [1, 100], `before` an ISO 8601 `createdAt` the page filters
// strictly before, and the client pages by passing the last row's `createdAt` as the
// next `before`. The parameter names `limit` and `before` are the API's own
// (apps/api/src/http/pagination.ts), which section 12 defers to. The wire carries no
// has-more flag, so iteration honestly ends on the first empty page.

@available(macOS 12.0, iOS 15.0, *)
final class PaginationTests: XCTestCase {
    /// A second games page, older than the fixture page's last row.
    private let olderGamesPage = Data(
        #"""
        {
          "games": [
            {
              "gameId": "d4e5f6a7-b8c9-4d0e-9f1a-3b4c5d6e7f8a",
              "name": null,
              "role": "solver",
              "createdAt": "2026-07-06T08:00:00.000Z",
              "createdBy": "a1b2c3d4-e5f6-4a7b-8c9d-0e1f2a3b4c5d",
              "memberCount": 2,
              "puzzle": {
                "puzzleId": "3f8c2b1a-9e4d-4f6a-b7c8-2d1e0f9a8b7c",
                "rows": 15,
                "cols": 15,
                "title": null
              }
            }
          ]
        }
        """#.utf8)

    func test_listGames_followsTheCreatedAtCursorAcrossTwoPagesAndEndsOnTheEmptyOne()
        async throws
    {
        let firstPage = try SharedRESTFixtures.data("games-list")
        let olderPage = olderGamesPage
        StubURLProtocol.install { request in
            switch request.queryValue("before") {
            case nil: return (200, firstPage)
            case "2026-07-07T09:30:00.000Z": return (200, olderPage)
            case "2026-07-06T08:00:00.000Z": return (200, Data(#"{"games":[]}"#.utf8))
            case let other: throw URLError(.unsupportedURL, userInfo: ["before": other ?? ""])
            }
        }
        let client = makeStubbedClient()

        let page1 = try await client.listGames(limit: 2)
        XCTAssertEqual(page1.rows.count, 2)
        XCTAssertEqual(
            page1.nextBefore, "2026-07-07T09:30:00.000Z",
            "the cursor is the server-computed nextBefore (the page-minimum createdAt)")

        // The older page predates activity ordering (no nextBefore key), so the client falls back
        // to the last row's createdAt: the two paths meet at the same cursor value here.
        let page2 = try await client.listGames(limit: 2, before: page1.nextBefore)
        XCTAssertEqual(page2.rows.count, 1)
        XCTAssertEqual(page2.rows[0].createdAt, "2026-07-06T08:00:00.000Z")
        XCTAssertEqual(page2.nextBefore, "2026-07-06T08:00:00.000Z")

        let page3 = try await client.listGames(limit: 2, before: page2.nextBefore)
        XCTAssertTrue(page3.rows.isEmpty)
        XCTAssertNil(page3.nextBefore, "an empty page ends iteration")

        let requests = StubURLProtocol.recordedRequests
        XCTAssertEqual(requests.count, 3)
        XCTAssertEqual(requests[0].queryValue("limit"), "2")
        XCTAssertNil(requests[0].queryValue("before"), "the first page sends no cursor")
        XCTAssertEqual(requests[1].queryValue("limit"), "2")
        XCTAssertEqual(requests[1].queryValue("before"), "2026-07-07T09:30:00.000Z")
        XCTAssertEqual(requests[2].queryValue("before"), "2026-07-06T08:00:00.000Z")
        for request in requests {
            XCTAssertEqual(request.path, "/games", "the cursor rides the query, not the path")
        }
    }

    func test_listPuzzles_passesTheCursorThroughAndEndsOnTheEmptyPage() async throws {
        let firstPage = try SharedRESTFixtures.data("puzzles-list")
        StubURLProtocol.install { request in
            if request.queryValue("before") == nil {
                return (200, firstPage)
            }
            return (200, Data(#"{"puzzles":[]}"#.utf8))
        }
        let client = makeStubbedClient()

        let page1 = try await client.listPuzzles()
        XCTAssertEqual(page1.rows.count, 2)
        XCTAssertEqual(page1.nextBefore, "2026-07-07T09:30:00.000Z")

        let page2 = try await client.listPuzzles(before: page1.nextBefore)
        XCTAssertTrue(page2.rows.isEmpty)
        XCTAssertNil(page2.nextBefore)

        let requests = StubURLProtocol.recordedRequests
        XCTAssertEqual(requests.count, 2)
        XCTAssertNil(requests[0].queryValue("before"))
        XCTAssertNil(requests[0].queryValue("limit"), "an unset limit is omitted, not defaulted")
        XCTAssertEqual(requests[1].queryValue("before"), "2026-07-07T09:30:00.000Z")
    }
}

File: apps/android/api/src/test/kotlin/crossy/api/PaginationTests.kt
// Cursor pagination for the two list endpoints (PROTOCOL.md §12): `limit` clamped server-side to
// [1, 100], `before` an ISO 8601 `createdAt` the page filters strictly before, and the client pages
// by the server-computed `nextBefore` (falling back to the last row's `createdAt` for an older
// server that omits it). The wire carries no has-more flag, so iteration honestly ends on the first
// empty page. Twin of apps/ios PaginationTests.swift.

package crossy.api

import kotlinx.coroutines.runBlocking
import okhttp3.mockwebserver.Dispatcher
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.RecordedRequest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class PaginationTests : MockServerTest() {

    @Test
    fun listGames_followsTheCursorAcrossTwoPagesAndEndsOnTheEmptyOne() = runBlocking {
        server.dispatcher = object : Dispatcher() {
            override fun dispatch(request: RecordedRequest): MockResponse =
                when (request.requestUrl?.queryParameter("before")) {
                    null -> jsonResponse(200, Fixtures.GAMES_LIST)
                    "2026-07-07T09:30:00.000Z" -> jsonResponse(200, Fixtures.GAMES_LIST_OLDER)
                    "2026-07-06T08:00:00.000Z" -> jsonResponse(200, Fixtures.GAMES_LIST_EMPTY)
                    else -> MockResponse().setResponseCode(400)
                }
        }
        val client = client()

        val page1 = client.listGames(limit = 2)
        assertEquals(2, page1.rows.size)
        assertEquals(
            "2026-07-07T09:30:00.000Z",
            page1.nextBefore,
            "the cursor is the server-computed nextBefore (the page-minimum createdAt)",
        )

        // The older page predates activity ordering (no nextBefore key), so the client falls back to
        // the last row's createdAt: the two paths meet at the same cursor value here.
        val page2 = client.listGames(limit = 2, before = page1.nextBefore)
        assertEquals(1, page2.rows.size)
        assertEquals("2026-07-06T08:00:00.000Z", page2.rows[0].createdAt)
        assertEquals("2026-07-06T08:00:00.000Z", page2.nextBefore)

        val page3 = client.listGames(limit = 2, before = page2.nextBefore)
        assertTrue(page3.rows.isEmpty())
        assertNull(page3.nextBefore, "an empty page ends iteration")

        assertEquals(3, server.requestCount)
        val first = server.takeRequest()
        assertEquals("/games", first.requestUrl?.encodedPath, "the cursor rides the query, not the path")
        assertEquals("2", first.requestUrl?.queryParameter("limit"))
        assertNull(first.requestUrl?.queryParameter("before"), "the first page sends no cursor")
        assertEquals("2026-07-07T09:30:00.000Z", server.takeRequest().requestUrl?.queryParameter("before"))
        assertEquals("2026-07-06T08:00:00.000Z", server.takeRequest().requestUrl?.queryParameter("before"))
    }

    @Test
    fun listGames_aPresentNullCursorMeansExhaustedNotFallback() = runBlocking {
        // Key present and null: the list is exhausted, and the client must NOT re-derive a cursor
        // from the reordered rows (§12: page by nextBefore, never by the visual last row).
        val exhausted = Fixtures.GAMES_LIST.replace(
            "\"nextBefore\": \"2026-07-07T09:30:00.000Z\"",
            "\"nextBefore\": null",
        )
        server.enqueue(jsonResponse(200, exhausted))

        val page = client().listGames()
        assertEquals(2, page.rows.size)
        assertNull(page.nextBefore, "a present-null cursor ends iteration on a full page")
    }

    @Test
    fun listPuzzles_passesTheCursorThroughAndEndsOnTheEmptyPage() = runBlocking {
        server.dispatcher = object : Dispatcher() {
            override fun dispatch(request: RecordedRequest): MockResponse =
                if (request.requestUrl?.queryParameter("before") == null) {
                    jsonResponse(200, Fixtures.PUZZLES_LIST)
                } else {
                    jsonResponse(200, Fixtures.PUZZLES_LIST_EMPTY)
                }
        }
        val client = client()

        val page1 = client.listPuzzles()
        assertEquals(2, page1.rows.size)
        assertEquals("2026-07-07T09:30:00.000Z", page1.nextBefore)

        val page2 = client.listPuzzles(before = page1.nextBefore)
        assertTrue(page2.rows.isEmpty())
        assertNull(page2.nextBefore)

        assertEquals(2, server.requestCount)
        val first = server.takeRequest()
        assertNull(first.requestUrl?.queryParameter("before"))
        assertNull(first.requestUrl?.queryParameter("limit"), "an unset limit is omitted, not defaulted")
        assertEquals("2026-07-07T09:30:00.000Z", server.takeRequest().requestUrl?.queryParameter("before"))
    }
}
